[DROOLS-7512] opt-in object store with references for reliable session

nprentza opened the pull request:
**JIRA**: 

[DROOLS-7512](https://issues.redhat.com/browse/DROOLS-7512)

Most of the SonarCloud issues raised were fixed in the last commit (sonar improvements).

tkobayas reviewed (approved):
Great job!

mariofusco commented on drools-reliability/drools-reliability-core/src/main/java/org/drools/reliability/core/SimpleSerializationReliableObjectStoreFactory.java:
```diff
@@ -25,6 +26,14 @@ public SimpleReliableObjectStore createSimpleReliableObjectStore(Storage<Long, S
     }
 
+    public SimpleReliableObjectStore createSimpleReliableObjectStore(Storage<Long, StoredObject> storage, PersistedSessionOption persistedSessionOption) {
+        if (persistedSessionOption.getPersistenceObjectsStrategy() == PersistedSessionOption.PersistenceObjectsStrategy.SIMPLE) {
```
It would be better to use a `switch` on the values of the `enum` here, so if for any reason we will have to add a third value we will immediately know that we will need to change also here.

nprentza replied:
yes, I will use a `switch`.

mariofusco commented on drools-reliability/drools-reliability-core/src/main/java/org/drools/reliability/core/SimpleSerializationReliableRefObjectStore.java:
```diff
@@ -0,0 +1,140 @@
+
+    private Map<String, Long> uniqueObjectTypesInStore;  // object type name, occurances
+
+    public SimpleSerializationReliableRefObjectStore() {
```
I don't understand why you added a constructor that should never be called here, can't it be simply removed?

nprentza replied:
Yes, you are right, you can remove it.

nprentza replied:
Sonar finds a bug in the class without the no-arguments constructor because its externalizable. Shall I put the constructor back?

mariofusco commented on drools-reliability/drools-reliability-core/src/main/java/org/drools/reliability/core/SimpleSerializationReliableRefObjectStore.java:
```diff
@@ -0,0 +1,140 @@
+
+    private void updateObjectTypesList() {
+        // list of unique object types in storage
+        List<String> uTypeNames = new ArrayList<>();
```
Maybe you could use a `Set` here instead of `List`, so you could avoid the check for uniqueness?

nprentza replied:
ok, I believe we can do that.

mariofusco commented on drools-reliability/drools-reliability-core/src/main/java/org/drools/reliability/core/SimpleSerializationReliableRefObjectStore.java:
```diff
@@ -0,0 +1,140 @@
+            }
+        });
+        // add unique object types + their occurrences in  uniqueObjectTypesInStore
+        uTypeNames.forEach(uType -> {
```
I believe you could achieve the same result in a clearer way using with a `collect` using `Collectors.groupingBy`.

nprentza replied:
here is how I used Collectors.qroupingBy, what do you think?

```
 uniqueObjectTypesInStore.putAll(storage.values().stream()
                .map(sObject -> sObject.getObject().getClass().getName())
                .collect(Collectors.groupingBy(Object::toString, Collectors.counting()))
                .entrySet().stream().filter(entry -> uTypeNames.contains((entry.getKey())))
                .collect(Collectors.toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue)));
```

mariofusco replied:
I believe (but I'm not 100% sure) that the following should be semantically equivalent, but easier to read

```
        uniqueObjectTypesInStore.putAll(storage.values().stream()
                .map(sObject -> sObject.getObject().getClass().getName())
                .filter(uTypeNames::contains)
                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting())));
```

Please let me know if I'm missing something.

nprentza replied:
my intention was to group first and then filter, for efficiency reasons, if it makes sense

mariofusco replied:
That's ok then, I'm not sure if it is more efficient to delay filter but having to recreate a second Map after that, or do everything in one shot as I suggested, but I guess there is no general answer and it depends case by case. Please fix the switch expression and I will merge this pull request.

nprentza replied:
yes, well for the moment maybe its better to have something that its easier to read. I will follow your suggestion and commit this change together with the switch fix.

mariofusco commented on drools-reliability/drools-reliability-core/src/main/java/org/drools/reliability/core/SimpleSerializationReliableRefObjectStore.java:
```diff
@@ -0,0 +1,140 @@
+    public SimpleSerializationReliableRefObjectStore(Storage<Long, StoredObject> storage) {
+        super(storage);
+        uniqueObjectTypesInStore = new HashMap<>();
+        if (storage.size() > 0) {
```
I find a bit confusing that you're repeating that `storage.size() > 0` check twice. Please do it only once simply assigning `this.storage` to `storage` in the else branch. Also it will be clearer to use `isEmpty()` instead of `size()`. In essence you could do something like:

```
if (storage.isEmpty()) {
    this.storage = storage;
} else {
    updateObjectTypesList();
    this.storage = updateObjectReferences(storage);
}
```

nprentza replied:
yes you are right, what you suggest is more clear.

mariofusco commented on drools-reliability/drools-reliability-core/src/main/java/org/drools/reliability/core/SimpleSerializationReliableRefObjectStore.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    private Long fromObjectToFactHandleId(Object object) {
+        for (Long key : this.storage.keySet()) {
```
This method is formally correct, but could be extremely inefficient. I'm not sure if feasible (please investigate) but maybe it would be better to add another map to this class, a `IdentityHashMap<Object, Long>` doing a reverse map from an object back to its key.

nprentza replied:
I will investigate and raise a separate PR for this.

nprentza commented:
> Nice job, I just suggested a few minor improvements. Please review them and we could have a final chat on this pull request during tomorrow's meeting before merging it.

Mario thanks a lot for your suggestions, I revised the code accordingly addressing all comments except the last one which I will address in a separate PR.

mariofusco commented on drools-reliability/drools-reliability-core/src/main/java/org/drools/reliability/core/SimpleSerializationReliableObjectStoreFactory.java:
```diff
@@ -25,6 +26,13 @@ public SimpleReliableObjectStore createSimpleReliableObjectStore(Storage<Long, S
     }
 
+    public SimpleReliableObjectStore createSimpleReliableObjectStore(Storage<Long, StoredObject> storage, PersistedSessionOption persistedSessionOption) {
+        switch (persistedSessionOption.getPersistenceObjectsStrategy()){
```
The whole point in using a `switch` statement here is having fast failure in case we add a new value in the enum and forget to manage it here. In essence I was thinking to something like the following, please let me know if it makes sense.

```
        switch (persistedSessionOption.getPersistenceObjectsStrategy()){
            case SIMPLE: return new SimpleSerializationReliableObjectStore(storage);
            case OBJECT_REFERENCES: return new SimpleSerializationReliableRefObjectStore(storage);
            default: throw new UnsupportedOperationException();
        }
```

nprentza replied:
Now I understand what you had in mind, yes, it make sense. Thanks.
